_posts/2016-07-11-using-lets-encrypt-for-a-multi-tenant-web-application.markdown

```diff
@@ -1,39 +1,60 @@
 
-### The ACME Protocol and DNS vs. HTTP Domain Validation
+* domain owner asks us to add their domain to our certificate
+* we send the request to the hosting provider and to the certificate authority
+* the certificate authority performs a domain validation process which is opaque to us
+* the certificate authority generates a new certificate
+* the certificate authority sends the certificate to the hosting provider
+* the hosting provider replaces our certificate with the new one and notifies us
+
+![previous certificate process](/img/posts/ssl_certificate_process_before.png)
+
+[Let's Encrypt](https://letsencrypt.org/) offers an automated process for obtaining and renewing SSL certificates -- including SAN SSL certificates for up to 100 domains -- as long as you can prove to them that you control any domain that's included in a requested certificate. Given that we wanted to be able to quickly offer HTTPS support to new or existing "masked" domain customers, this automation seemed very promising to us. The fact that Let's Encrypt is a free service made it all the more compelling. So we set ourselves about the task of making sure we could always prove our ownership of the `search.usa.gov` domain as well as "masked" customer domains while keeping the site available for customer use at all times.
+
+## The ACME Protocol and DNS vs. HTTP Domain Validation
 
 The process of requesting an SSL certificate from Let's Encrypt involves running a Let's Encrypt agent locally which communicates with the Let's Encrypt Certificate Authority (CA) using the [ACME Protocol](https://github.com/letsencrypt/acme-spec). The CA issues challenges that the agent attempts to respond to, and the challenges come in the user's choice of two forms:
 
 * DNS challenge: the agent creates a DNS record for a particular sub-domain requested by the CA in order to prove that it has control over DNS for that domain
 * HTTP challenge: the agent creates a text file at a particular location with particular content that the CA can then request in order to prove that the agent has control over web page content for that domain
 
-Since we didn't have DNS control over `search.usa.gov`, that left HTTP as the only Domain Validation option for us. Our goal then was to make sure we could have the Let's Encrypt agent create web resources as needed to respond to any HTTP challenge issued by the CA.
+Since we didn't have an easy way to publish `search.usa.gov` DNS records, that left HTTP as the only Domain Validation option for us. Our goal then was to make sure we could have the Let's Encrypt agent create web resources as needed to respond to any HTTP challenge issued by the CA.
```

Comment:
I realized that I misstated the real issue here when we spoke this week.

Since we'd be validating our _customers'_ domains, it doesn't matter if we have an easy way to public DNS records in `search.usa.gov` because we don't have any way at all to publish DNS records in their domains! That's really the best way to phrase this, and it's pertinent to the idea of a multi-tenant SaaS with domain masking.